AD array variables/kernels for finite elements

lindsayad opened the pull request:
Closes #31506

lindsayad committed:
Add FE AD array variable support

Also removes unnecessary scoped locks because I was bored and
we don't need them any more since we lock at the PetscVector
and PetscMatrix levels

Refs #31506

Co-authored-by: Oana Marin <oana_marin@outlook.com>

lindsayad committed:
Add ADArrayNodalKernel

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3324302/), step Results summary on 2014aae wanted to post the following:

## Framework test summary

### Removed tests
None
### Added tests
| Test                                                    |   Time (s) |
|---------------------------------------------------------|------------|
| `kernels/array_kernels.test_ad_diffusion_reaction_jac`  |       1.08 |
| `kernels/array_kernels.test_ad_jac`                     |       0.89 |
| `nodalkernels/array-reaction-decay.test/result_diff`    |       0.86 |
| `kernels/array_kernels.test_ad`                         |       0.85 |
| `nodalkernels/array-reaction-decay.test/jac`            |       0.85 |
| `nodalkernels/array-reaction-decay.test/ad_jac`         |       0.82 |
| `nodalkernels/array-reaction-decay.test/ad_result_diff` |       0.81 |
| `kernels/array_kernels.test_ad_diffusion_reaction`      |       0.8  |
### Run time changes
| Test                                                                         |   Base (s) |   Head (s) | +/-       |
|------------------------------------------------------------------------------|------------|------------|-----------|
| `meshdivisions.mesh_div/extra_id`                                            |       0.93 |      39.37 | +4141.17% |
| `meshdivisions.mesh_div/nested`                                              |       1.12 |      39.5  | +3420.39% |
| `geomsearch/3d_penetration_locator.rings_facing_each_other`                  |       2.24 |       8.9  | +297.42%  |
| `variables/get_elemental_value.get_elemental_val/old`                        |       0.88 |       2.18 | +147.35%  |
| `variables/get_elemental_value.get_elemental_val/older`                      |       0.88 |       2.15 | +144.01%  |
| `variables/get_elemental_value.get_elemental_val/current`                    |       0.81 |       1.97 | +143.51%  |
| `meshdivisions.mesh_div/cartesian`                                           |       0.87 |       1.94 | +123.54%  |
| `meshdivisions.mesh_div/spherical`                                           |       0.81 |       1.73 | +113.37%  |
| `meshdivisions.mesh_div/cartesian_center_pos`                                |       0.77 |       1.65 | +112.59%  |
| `meshdivisions.mesh_div/positions`                                           |       0.84 |       1.76 | +108.52%  |
| `meshdivisions.mesh_div/cartesian_center`                                    |       0.82 |       1.7  | +106.69%  |
| `meshdivisions.mesh_div/functor`                                             |       0.81 |       1.62 | +100.58%  |
| `meshdivisions.mesh_div/cylinder`                                            |       0.82 |       1.61 | +95.46%   |
| `outputs/debug.show_execution_order/user_objects_transfers`                  |       1.26 |       2.46 | +94.84%   |
| `meshdivisions.mesh_div/cylinder_center_positions`                           |       0.84 |       1.62 | +92.39%   |
| `meshdivisions.mesh_div/spherical_center_positions`                          |       0.85 |       1.62 | +89.88%   |
| `meshgenerators/transfinite_generator.generate/annular`                      |       0.76 |       1.3  | +69.75%   |
| `meshgenerators/annular_mesh_generator.annular_mesh_generator_deprecated`    |       0.79 |       1.33 | +66.85%   |
| `meshgenerators/block_deletion_generator.delete/block_deletion_test3`        |       0.62 |       1.01 | +63.41%   |
| `functions/parsed.vector`                                                    |       0.92 |       1.47 | +59.22%   |
| `transfers/multiapp_postprocessor_interpolation_transfer.basis/radial_basis` |       1.01 |       1.57 | +55.63%   |
## Modules test summary

### Removed tests
None
### Added tests
None
### Run time changes
| Test                                                                                                       |   Base (s) |   Head (s) | +/-        |
|------------------------------------------------------------------------------------------------------------|------------|------------|------------|
| `peridynamics/test:jacobian_check.2D_thermomechanics_free_node_BPD`                                        |       0.81 |     299.82 | +37089.47% |
| `peridynamics/test:simple_tests.2D_regularD_variableH_OSPD`                                                |       0.91 |     299.87 | +32793.59% |
| `peridynamics/test:restart.initial_run`                                                                    |       0.91 |     299.97 | +32747.34% |
| `peridynamics/test:heat_conduction.2D_steady_state_BPD`                                                    |       0.92 |     299.96 | +32589.67% |
| `peridynamics/test:simple_tests.2D_irregularD_variableH_BPD`                                               |       0.92 |     299.78 | +32457.14% |
| `peridynamics/test:simple_tests.2D_regularD_constH_OSPD`                                                   |       0.93 |     299.9  | +32150.25% |
| `peridynamics/test:jacobian_check.2D_mechanics_free_node_OSPD`                                             |       0.94 |     299.81 | +31727.90% |
| `peridynamics/test:generalized_plane_strain.out_of_plane_pressure_OSPD`                                    |       0.96 |     299.93 | +31286.56% |
| `peridynamics/test:jacobian_check.2D_thermomechanics_BPD`                                                  |       0.96 |     299.81 | +31172.75% |
| `peridynamics/test:jacobian_check.2D_mechanics_OSPD`                                                       |       0.98 |     299.81 | +30427.27% |
| `peridynamics/test:nodalkernels.test`                                                                      |       0.98 |     299.98 | +30385.79% |
| `peridynamics/test:generalized_plane_strain.gps_OSPD`                                                      |       1    |     299.98 | +30038.90% |
| `thermal_hydraulics/test:problems/sod_shock_tube.davis_wave_speed`                                         |       1.06 |     299.93 | +28114.70% |
| `peridynamics/test:generalized_plane_strain.planestrain_prescribed_OSPD`                                   |       1.09 |     299.98 | +27401.75% |
| `thermal_hydraulics/test:problems/sod_shock_tube.non_x_direction_1phase`                                   |       1.11 |     300.25 | +27027.38% |
| `peridynamics/test:generalized_plane_strain.gps_H1NOSPD`                                                   |       1.11 |     299.84 | +26968.49% |
| `thermal_hydraulics/test:problems/double_rarefaction.1phase`                                               |       1.11 |     299.95 | +26854.69% |
| `thermal_hydraulics/test:problems/sod_shock_tube.sod_shock_tube_erk2_light`                                |       1.13 |     299.97 | +26525.67% |
| `peridynamics/test:failure_tests.2D_stretch_failure_BPD`                                                   |       1.13 |     299.95 | +26504.64% |
| `thermal_hydraulics/test:problems/lax_shock_tube.all/explicit`                                             |       1.14 |     300.01 | +26132.85% |
| `rdg/test:advection_1d.rdgP0`                                                                              |       0.93 |     219.39 | +23517.81% |
| `peridynamics/test:plane_stress.conventional_planestress_OSPD`                                             |       1.28 |     300.01 | +23302.68% |
| `thermal_hydraulics/test:problems/sod_shock_tube.sod_shock_tube_be`                                        |       1.33 |     299.93 | +22382.15% |
| `peridynamics/test:simple_tests.2D_irregularD_variableH_OSPD`                                              |       1.35 |     299.7  | +22120.57% |
| `thermal_hydraulics/tutorials/single_phase_flow.tutorial/01`                                               |       1.36 |     299.98 | +21902.42% |
| `peridynamics/test:plane_stress.conventional_planestress_H1NOSPD`                                          |       1.37 |     299.98 | +21860.66% |
| `peridynamics/test:failure_tests.2D_bond_status_convergence_BPD`                                           |       1.38 |     299.98 | +21672.68% |
| `peridynamics/test:failure_tests.2D_singular_shape_tensor_H1NOSPD`                                         |       1.41 |     299.93 | +21222.82% |
| `peridynamics/test:plane_stress.weak_planestress_H1NOSPD`                                                  |       1.41 |     299.89 | +21157.18% |
| `peridynamics/test:auxkernels.stretch_H1NOSPD`                                                             |       1.42 |     300.19 | +21082.45% |
| `peridynamics/test:auxkernels.offset_and_volume_3D`                                                        |       0.99 |     209.28 | +21026.65% |
| `thermal_hydraulics/test:problems/square_wave.square_wave_minmod`                                          |       1.44 |     299.98 | +20725.71% |
| `thermal_hydraulics/test:problems/square_wave.square_wave_mc`                                              |       1.45 |     299.91 | +20616.72% |
| `peridynamics/test:jacobian_check.2D_mechanics_BPD`                                                        |       0.88 |     173.92 | +19711.22% |
| `peridynamics/test:simple_tests.2D_small_strain_H1NOSPD`                                                   |       1.54 |     299.8  | +19306.30% |
| `peridynamics/test:auxkernels.ranktwotensor_OSPD`                                                          |       1.56 |     299.98 | +19125.92% |
| `thermal_hydraulics/tutorials/single_phase_flow.tutorial/02`                                               |       1.57 |     299.91 | +19040.11% |
| `thermal_hydraulics/test:problems/lax_shock_tube.all/implicit`                                             |       1.61 |     299.9  | +18534.07% |
| `thermal_hydraulics/test:problems/square_wave.square_wave_superbee`                                        |       1.64 |     299.88 | +18192.41% |
| `peridynamics/test:jacobian_check.2D_mechanics_free_node_BPD`                                              |       0.89 |     158.6  | +17641.40% |
| `thermal_hydraulics/test:components/simple_turbine_1phase.phy:conservation`                                |       1.7  |     299.98 | +17564.59% |
| `thermal_hydraulics/test:components/volume_junction_1phase.conservation_1phase`                            |       1.7  |     299.89 | +17561.30% |
| `peridynamics/test:generalized_plane_strain.gps_squares_OSPD`                                              |       1.71 |     300.06 | +17433.34% |
| `peridynamics/test:failure_tests.2D_stress_failure_H1NOSPD`                                                |       1.94 |     300.03 | +15402.52% |
| `thermal_hydraulics/test:problems/pressure_drop.tests/without_junction`                                    |       2.36 |     299.97 | +12619.18% |
| `thermal_hydraulics/tutorials/single_phase_flow.tutorial/03`                                               |       2.61 |     299.9  | +11392.82% |
| `thermal_hydraulics/test:problems/pressure_drop.tests/with_junction`                                       |       2.95 |     299.9  | +10053.63% |
| `peridynamics/test:simple_tests.2D_regularD_variableH_BPD`                                                 |       0.86 |      73.95 | +8539.69%  |
| `peridynamics/test:simple_tests.2D_regularD_constH_BPD`                                                    |       0.9  |      73.24 | +8062.95%  |
| `thermal_hydraulics/test:problems/brayton_cycle.closed`                                                    |       4.21 |     299.93 | +7016.21%  |
| `thermal_hydraulics/tutorials/single_phase_flow.tutorial/04`                                               |       4.28 |     299.86 | +6906.45%  |
| `rdg/test:advection_1d.1d_aefv_square_wave_minmod`                                                         |       0.82 |      48.32 | +5802.39%  |
| `rdg/test:advection_1d.1d_aefv_square_wave_none`                                                           |       0.83 |      48.63 | +5773.06%  |
| `peridynamics/test:jacobian_check.2D_heat_conduction_BPD`                                                  |       0.83 |      47.36 | +5593.72%  |
| `rdg/test:advection_1d.1d_aefv_square_wave_mc`                                                             |       0.85 |      48.25 | +5591.03%  |
| `peridynamics/test:jacobian_check.2D_heat_conduction_free_node_BPD`                                        |       0.87 |      47.57 | +5388.52%  |
| `thermal_hydraulics/tutorials/single_phase_flow.tutorial/05`                                               |       5.48 |     299.79 | +5374.25%  |
| `rdg/test:advection_1d.1d_aefv_square_wave_superbee`                                                       |       0.85 |      44.28 | +5107.53%  |
| `thermal_hydraulics/test:problems/brayton_cycle.open`                                                      |       4.19 |     213.05 | +4979.66%  |
| `combined/test:optimization/optimization_density_update.top_opt_2d_pde_filter`                             |       6.06 |     300.03 | +4847.82%  |
| `combined/test:optimization/optimization_density_update.top_opt_2d`                                        |       6.12 |     300.01 | +4799.61%  |
| `peridynamics/test:auxkernels.offset_and_area_2D`                                                          |       0.94 |      43.79 | +4539.63%  |
| `peridynamics/test:simple_tests.2D_finite_strain_H1NOSPD`                                                  |       1.57 |      70.14 | +4375.35%  |
| `thermal_hydraulics/test:misc/uniform_refine.test`                                                         |       1.03 |      44.49 | +4223.97%  |
| `combined/test:optimization/thermal_sensitivity.thermal_compliance/2d_root`                                |       2.44 |     101.86 | +4073.11%  |
| `optimization/test:simp.simp`                                                                              |       1.55 |      59.3  | +3735.75%  |
| `optimization/test:simp.simp_restricted`                                                                   |       1.71 |      59.6  | +3379.06%  |
| `thermal_hydraulics/test:problems/area_constriction.with_junction`                                         |       3.6  |     115.2  | +3102.82%  |
| `reactor/test:meshdivisions.mesh_div/single_hex`                                                           |       0.91 |      25.36 | +2689.47%  |
| `rdg/test:advection_1d.1d_block_restrictable`                                                              |       0.87 |      22.17 | +2435.57%  |
| `combined/test:subchannel_thm_coupling.THM`                                                                |      12.22 |     299.9  | +2353.29%  |
| `reactor/test:meshdivisions.mesh_div/distributed_hexes`                                                    |       0.82 |      16.44 | +1893.02%  |
| `thermal_hydraulics/test:components/junction_one_to_one_1phase.phy:unequal_area_1phase`                    |       1.53 |      25.11 | +1540.80%  |
| `combined/test:subchannel_thm_coupling.THM_pump`                                                           |      22.93 |     299.94 | +1207.88%  |
| `thermal_hydraulics/test:components/junction_one_to_one_1phase.phy:no_junction_equivalency_1phase_flipped` |       1.3  |      12.93 | +894.16%   |
| `thermal_hydraulics/test:components/junction_one_to_one_1phase.phy:no_junction_equivalency_1phase_normal`  |       1.31 |      12.8  | +878.90%   |
| `optimization/test:simp.simp_multi_restricted`                                                             |       3.14 |      24.24 | +671.03%   |
| `optimization/test:simp.simp_multi`                                                                        |       3.88 |      27.68 | +613.03%   |
| `combined/test:optimization/compliance_sensitivity.compliance_sensitivity/3d_mbb`                          |      15.77 |      94.18 | +497.13%   |
| `thermal_hydraulics/test:problems/freefall.test`                                                           |       1.05 |       6.17 | +485.24%   |
| `combined/test:optimization/compliance_sensitivity.thermal_test`                                           |       4.84 |      21    | +333.75%   |
| `peridynamics/test:jacobian_check.2D_thermomechanics_OSPD`                                                 |       1.09 |       3.1  | +184.11%   |
| `porous_flow/test:fluidstate.geothermal_1d_radial`                                                         |       2.06 |       4.94 | +139.67%   |
| `porous_flow/test:fluidstate.geothermal_1d`                                                                |       1.73 |       3.76 | +117.19%   |
| `porous_flow/examples/co2_intercomparison/1Dradial.brine_co2`                                              |       0.89 |       1.44 | +62.98%    |
| `porous_flow/examples/co2_intercomparison/1Dradial.water_and_co2`                                          |       0.91 |       1.39 | +52.62%    |

lindsayad committed:
Add doc

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3326041/), step Results summary on 4353e0c wanted to post the following:

## Framework test summary

### Removed tests
None
### Added tests
| Test                                                    |   Time (s) |
|---------------------------------------------------------|------------|
| `nodalkernels/array-reaction-decay.test/ad_result_diff` |       1.26 |
| `kernels/array_kernels.test_ad_jac`                     |       1.13 |
| `kernels/array_kernels.test_ad_diffusion_reaction_jac`  |       1.11 |
| `nodalkernels/array-reaction-decay.test/jac`            |       1    |
| `nodalkernels/array-reaction-decay.test/ad_jac`         |       0.99 |
| `kernels/array_kernels.test_ad`                         |       0.98 |
| `nodalkernels/array-reaction-decay.test/result_diff`    |       0.97 |
| `kernels/array_kernels.test_ad_diffusion_reaction`      |       0.95 |
### Run time changes
| Test                                                                                             |   Base (s) |   Head (s) | +/-     |
|--------------------------------------------------------------------------------------------------|------------|------------|---------|
| `outputs/misc.getOutputs`                                                                        |       1.28 |       2.22 | +74.03% |
| `meshgenerators/sidesets_between_subdomains_generator.group/nonadjacent_subs`                    |       0.79 |       1.34 | +70.19% |
| `geomsearch/2d_moving_penetration.pl_test1`                                                      |       0.85 |       1.4  | +64.04% |
| `meshgenerators/break_mesh_by_block_generator.surrounding_block_restricted/split_interface_only` |       0.83 |       1.32 | +58.98% |
| `meshgenerators/fill_between_point_vectors_generator.bow`                                        |       0.9  |       1.42 | +57.48% |
| `restep.no_restep/transient_no_solve`                                                            |       0.83 |       1.29 | +56.29% |
| `meshgenerators/transfinite_generator.generate/parsed_annular`                                   |       0.75 |       1.17 | +56.09% |
| `mesh/face_info.face_info/tri`                                                                   |       0.86 |       1.34 | +55.75% |
| `meshgenerators/move_node_generator.generate/3d`                                                 |       0.9  |       1.4  | +55.65% |
| `misc/jacobian.zero_jacobian_2D_ok`                                                              |       0.83 |       1.29 | +55.46% |
| `kernels/conservative_advection.upwinding/no_upwinding_1D`                                       |       0.89 |       1.39 | +55.31% |
| `physics.cg_ad`                                                                                  |       0.98 |       1.51 | +54.35% |
| `postprocessors/side_integral.functors/domain_side`                                              |       0.97 |       1.49 | +54.13% |
| `transfers/multiapp_scalar_to_auxscalar_transfer/to_sub.parent_to_sub`                           |       0.91 |       1.39 | +52.20% |
| `meshgenerators/parsed_generate_sideset.missing_errors/invalid_included_boundary`                |       1.21 |       1.82 | +50.86% |
| `vectorpostprocessors/least_squares_fit.least_squares_csv0`                                      |       0.8  |       1.2  | +50.06% |
## Modules test summary

### Removed tests
None
### Added tests
None
### Run time changes
| Test                                                                              |   Base (s) |   Head (s) | +/-     |
|-----------------------------------------------------------------------------------|------------|------------|---------|
| `richards/test:uo_egs.seff2`                                                      |       0.82 |       1.37 | +66.40% |
| `porous_flow/examples/multiapp_fracture_flow/fracture_diffusion.conforming_dirac` |       0.72 |       1.13 | +55.92% |
| `solid_mechanics/tutorials/introduction.mech_step04a`                             |       0.82 |       1.25 | +52.28% |

lindsayad committed:
More variable refactoring

lindsayad committed:
Fix doc paths to avoid ambiguity

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3329466/), step Results summary on 59bb993 wanted to post the following:

## Framework test summary

Compared against 2bd28b1 in job [civet.inl.gov/job/3326314](https://civet.inl.gov/job/3326314).

### Removed tests



### Added tests

| Test                                                    | Time (s)   |
|---------------------------------------------------------|------------|
| `nodalkernels/array-reaction-decay.test/ad_jac`         | 1.20       |
| `kernels/array_kernels.test_ad_jac`                     | 1.13       |
| `nodalkernels/array-reaction-decay.test/jac`            | 1.00       |
| `kernels/array_kernels.test_ad_diffusion_reaction_jac`  | 0.92       |
| `kernels/array_kernels.test_ad_diffusion_reaction`      | 0.91       |
| `nodalkernels/array-reaction-decay.test/ad_result_diff` | 0.90       |
| `kernels/array_kernels.test_ad`                         | 0.87       |
| `nodalkernels/array-reaction-decay.test/result_diff`    | 0.86       |

### Run time changes
## Modules test summary

Compared against 2bd28b1 in job [civet.inl.gov/job/3326314](https://civet.inl.gov/job/3326314).

### Removed tests



### Added tests



### Run time changes

lindsayad committed:
locks still needed for nodal normals uos

We have a read-write race here in PetscVector for the `_read_only_values`
array. There are locks for access to this array in both `_restore_array`
and `_get_array` but `operator()` which first calls `_get_array` then
tries to read from `_read_only_values` without a lock, while a method
like `set` may already be operating on `_read_only_values` (with a lock)
in `_restore_array`. I don't know what a fix here would be

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3335171/), step Results summary on d51e6b4 wanted to post the following:

## Framework test summary

Compared against ccb56d5 in job [civet.inl.gov/job/3330324](https://civet.inl.gov/job/3330324).

### Removed tests



### Added tests

| Test                                                    | Time (s)   |
|---------------------------------------------------------|------------|
| `kernels/array_kernels.test_ad_diffusion_reaction`      | 1.00       |
| `kernels/array_kernels.test_ad_jac`                     | 0.90       |
| `nodalkernels/array-reaction-decay.test/ad_result_diff` | 0.89       |
| `kernels/array_kernels.test_ad_diffusion_reaction_jac`  | 0.89       |
| `nodalkernels/array-reaction-decay.test/jac`            | 0.88       |
| `nodalkernels/array-reaction-decay.test/ad_jac`         | 0.87       |
| `kernels/array_kernels.test_ad`                         | 0.87       |
| `nodalkernels/array-reaction-decay.test/result_diff`    | 0.85       |

### Run time changes
## Modules test summary

Compared against ccb56d5 in job [civet.inl.gov/job/3330324](https://civet.inl.gov/job/3330324).

### Removed tests



### Added tests



### Run time changes

| Test                                                                        | Base (s)   | Head (s)   | +/-     |
|-----------------------------------------------------------------------------|------------|------------|---------|
| `stochastic_tools/test:web_server_control.stochastic_control/batch_restore` | 2.50       | 3.90       | +56.05% |

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3351996/), step Results summary on a909af8 wanted to post the following:

## Framework test summary

Compared against 8593683 in job [civet.inl.gov/job/3351776](https://civet.inl.gov/job/3351776).

### Removed tests



### Added tests

| Test                                                    | Time (s)   |
|---------------------------------------------------------|------------|
| `kernels/array_kernels.test_ad_diffusion_reaction_jac`  | 1.08       |
| `kernels/array_kernels.test_ad_jac`                     | 0.94       |
| `nodalkernels/array-reaction-decay.test/ad_result_diff` | 0.91       |
| `nodalkernels/array-reaction-decay.test/ad_jac`         | 0.87       |
| `kernels/array_kernels.test_ad`                         | 0.82       |
| `kernels/array_kernels.test_ad_diffusion_reaction`      | 0.80       |

### Run time changes
## Modules test summary

Compared against 8593683 in job [civet.inl.gov/job/3351776](https://civet.inl.gov/job/3351776).

### Removed tests



### Added tests



### Run time changes

| Test                                                                              | Base (s)   | Head (s)   | +/-     |
|-----------------------------------------------------------------------------------|------------|------------|---------|
| `stochastic_tools/test:web_server_control.stochastic_control/batch_restore_multi` | 2.74       | 4.26       | +55.40% |

lindsayad commented:
griffin patch at https://github.inl.gov/ncrc/griffin/pull/2812

lindsayad commented:
@bwspenc I dunno if you want to review this one too?

GiudGiud commented on framework/include/bcs/ADArrayDirichletBC.h:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Boundary condition of a Dirichlet type
+ *
+ * Sets the value in the node
```
```suggestion
 * Sets the value at the node for each component of the array variable
```

GiudGiud commented on framework/src/bcs/ADArrayDirichletBC.C:
```diff
@@ -0,0 +1,41 @@
+ADRealEigenVector
+ADArrayDirichletBC::computeQpResidual()
+{
+  return _u - _values;
```
so we enforce this one strongly

do we have a place in the doco where we say we enforce DirichletBCs strongly ?

GiudGiud commented on framework/include/kernels/ADArrayKernel.h:
```diff
@@ -0,0 +1,77 @@
+#include "MooseVariableInterface.h"
+#include "ADFunctorInterface.h"
+
+class ADArrayKernel : public KernelBase,
```
```suggestion
/**
 * Base class for array variable (equation) kernels using automatic differentiation
 */
class ADArrayKernel : public KernelBase,
```

GiudGiud commented on framework/include/kernels/ADArrayReaction.h:
```diff
@@ -0,0 +1,30 @@
+  const MaterialProperty<Real> * _r;
+  /// array diffusion coefficient
+  const MaterialProperty<RealEigenVector> * const _r_array;
+  /// matrix diffusion coefficient
```
```suggestion
  /// matrix diffusion coefficient (notably for anisotropic diffusion)
```

GiudGiud commented on framework/include/variables/MooseVariableData.h:
```diff
@@ -425,22 +428,22 @@ class MooseVariableData : public MooseVariableDataBase<OutputType>
+  const DofValues & dofValuesDot() const;
+  const DofValues & dofValuesDotOld() const;
+  const DofValues & dofValuesDotDot() const;
+  const DofValues & dofValuesDotDotOld() const;
```
ok you did have to change it in a bunch of places

GiudGiud commented on framework/include/variables/MooseVariableFE.h:
```diff
@@ -29,17 +29,19 @@ class MooseVariableFE;
 typedef MooseVariableFE<Real> MooseVariable;
 typedef MooseVariableFE<RealVectorValue> VectorMooseVariable;
 typedef MooseVariableFE<RealEigenVector> ArrayMooseVariable;
+// typedef MooseVariableFE<ADRealEigenVector> ADArrayMooseVariable;
```
not sure why we want a special type? in therms of AD derivative values, it's still just "1" for the relevant dof and 0 for all others?

GiudGiud commented on framework/src/kernels/ADArrayReaction.C:
```diff
@@ -0,0 +1,55 @@
+
+ADArrayReaction::ADArrayReaction(const InputParameters & parameters)
+  : ADArrayKernel(parameters),
+    _r(hasMaterialProperty<Real>("reaction_coefficient")
```
```suggestion
    _r(hasMaterialProperty<ADReal>("reaction_coefficient")
```
same for the others

ADArrayReaction should be ADArrayMatReaction and support AD everywhere

GiudGiud commented on framework/src/nodalkernels/ADArrayNodalKernel.C:
```diff
@@ -0,0 +1,68 @@
+ADArrayNodalKernel::computeResidual()
+{
+  if (!_var.isNodalDefined())
+    return;
```
should we error in constructor then? why allow a non-nodal variable as the variable for a nodal kernel?

GiudGiud replied:
ah my b

GiudGiud commented on framework/src/variables/MooseVariableDataBase.C:
```diff
@@ -522,6 +508,7 @@ MooseVariableDataBase<OutputType>::nodalValueArray(Moose::SolutionState state) c
       default:
         mooseError("No current support for PreviousNL for nodal value array");
     }
+    Moose::out << "_nodal_value_array.size() : " << _nodal_value_array.size() << std::endl;
```
```suggestion
```

GiudGiud commented on framework/src/variables/MooseVariableDataBase.C:
```diff
@@ -446,22 +439,15 @@ MooseVariableDataBase<RealEigenVector>::add(NumericVector<Number> & residual)
+    mooseAssert(_dof_indices.size() % _count == 0,
+                "Dof indices should be cleanly divisible by the variable count");
+    const auto n_shapes = _dof_indices.size() / _count;
+    for (const auto indx : index_range(_dof_indices))
```
we're losing the potential to vectorize over the vector in the nodal case. we had a clear access pattern and now we're accessing inside a vector to get the index

lindsayad replied:
The thing is that the old code had zero robustness. Even for a `isNodal()` `LAGRANGE_VEC` variable this would not have worked because the old `_dof_indices` behavior assumed a stride of 1 between variable components in the global dof index numbering but with a `LAGRANGE_VEC` that stride is no longer 1. We may lose some performance here but I much prefer robustness over speed

lindsayad replied:
the stride between the same shape function for different array variable components is the number of shapes for a single component, e.g. `n_shapes`

lindsayad commented on framework/include/utils/MooseTypes.h:
```diff
@@ -378,6 +412,9 @@ typedef MooseArray<ADRealVectorValue> ADVariableGradient;
 typedef MooseArray<ADRealTensorValue> ADVariableSecond;
 typedef MooseArray<ADRealVectorValue> ADVectorVariableValue;
 typedef MooseArray<ADRealTensorValue> ADVectorVariableGradient;
+// typedef MooseArray<ADRealEigenVector> ADArrayVariableValue;
```
I must have this somewhere else because I use it in `ADArrayKernel.h`

lindsayad commented on framework/src/nodalkernels/ArrayNodalKernel.C:
```diff
@@ -29,41 +29,33 @@ ArrayNodalKernel::ArrayNodalKernel(const InputParameters & parameters)
-{
-  const dof_id_type root_dof_idx = _var.nodalDofIndex();
-  std::iota(_work_dofs.begin(), _work_dofs.end(), root_dof_idx);
-}
```
This is no longer needed because libMesh now gives us the complete array dof indices without us needing to do anything in MOOSE

lindsayad committed:
Apply suggestions from code review

- Add doxygen
- Typo fixes
- Remove some commented code

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad committed:
Set nl_abs_tol to avoid some test harness failures

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3361040/), step Results summary on 14dff08 wanted to post the following:

## Framework test summary

Compared against 0cc44f8 in job [civet.inl.gov/job/3354977](https://civet.inl.gov/job/3354977).

### Removed tests



### Added tests

| Test                                                    | Time (s)   |
|---------------------------------------------------------|------------|
| `nodalkernels/array-reaction-decay.test/ad_jac`         | 0.98       |
| `kernels/array_kernels.test_ad_diffusion_reaction_jac`  | 0.94       |
| `kernels/array_kernels.test_ad_jac`                     | 0.92       |
| `kernels/array_kernels.test_ad`                         | 0.83       |
| `kernels/array_kernels.test_ad_diffusion_reaction`      | 0.82       |
| `nodalkernels/array-reaction-decay.test/ad_result_diff` | 0.82       |

### Run time changes

| Test                                              | Base (s)   | Head (s)   | +/-     |
|---------------------------------------------------|------------|------------|---------|
| `fvkernels/mms/advective-outflow.outflow/average` | 28.62      | 7.87       | -72.50% |
## Modules test summary

Compared against 0cc44f8 in job [civet.inl.gov/job/3354977](https://civet.inl.gov/job/3354977).

### Removed tests



### Added tests



### Run time changes

GiudGiud commented on framework/src/kernels/ADArrayReaction.C:
```diff
@@ -0,0 +1,55 @@
+ADArrayReaction::ADArrayReaction(const InputParameters & parameters)
+  : ADArrayKernel(parameters),
+    _r(hasMaterialProperty<ADReal>("reaction_coefficient")
+           ? &getMaterialProperty<Real>("reaction_coefficient")
```
```suggestion
           ? &getMaterialProperty<ADReal>("reaction_coefficient")
```
should be the same for all types imo

GiudGiud commented on framework/src/bcs/ADArrayDirichletBC.C:
```diff
@@ -0,0 +1,41 @@
+ADArrayDirichletBC::validParams()
+{
+  InputParameters params = ADArrayNodalBC::validParams();
+  params.addRequiredParam<RealEigenVector>("values",
```
you d probably need a test.
I remeber vectors being referred to in controllable code, but I have never tested it either.

lindsayad committed:
Address Guillaume review round 2

- Actually convert all the reaction material types to AD
- Required generalizing dataStore/dataLoad for Eigen::Matrix

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3363769/), step Results summary on 0b47451 wanted to post the following:

## Framework test summary

Compared against fe4ab42 in job [civet.inl.gov/job/3359230](https://civet.inl.gov/job/3359230).

### Removed tests



### Added tests

| Test                                                    | Time (s)   |
|---------------------------------------------------------|------------|
| `kernels/array_kernels.test_ad_diffusion_reaction`      | 1.27       |
| `kernels/array_kernels.test_ad_diffusion_reaction_jac`  | 1.26       |
| `kernels/array_kernels.test_ad_jac`                     | 1.11       |
| `nodalkernels/array-reaction-decay.test/ad_jac`         | 0.92       |
| `nodalkernels/array-reaction-decay.test/ad_result_diff` | 0.86       |
| `kernels/array_kernels.test_ad`                         | 0.84       |

### Run time changes

| Test                                              | Base (s)   | Head (s)   | +/-     |
|---------------------------------------------------|------------|------------|---------|
| `fvkernels/mms/advective-outflow.outflow/average` | 28.40      | 9.02       | -68.22% |
## Modules test summary

Compared against fe4ab42 in job [civet.inl.gov/job/3359230](https://civet.inl.gov/job/3359230).

### Removed tests



### Added tests



### Run time changes

bwspenc commented on framework/doc/content/source/kernels/ArrayDiffusion.md:
```diff
@@ -26,7 +26,7 @@ We can rearrange the fully expanded [eq:weak-array-diffusion] into
 (\nabla \vec{u}^\ast, \mathbf{D} \nabla \vec{u}) = \sum_{e} \sum_{i=1}^{N_{\text{dof}}} \sum_{\text{qp}=1}^{N_{qp}} (|J|w)_{\text{qp}} \vec{w}_p\vec{u}_i^\ast \underline{\mathbf{D}_{\text{qp}} \nabla \vec{u}_{\text{qp}} \cdot \nabla b_{i,\text{qp}}},
 \end{equation}
-where the underlined term is the array provided by [ArrayDiffusion::computeQpResidual](ArrayDiffusion.C).
+where the underlined term is the array provided by [ArrayDiffusion::computeQpResidual](/ArrayDiffusion.C).
```
Just curious about what the purpose of the leading slash is. Is that to disambiguate ADArrayDiffusion.C and ArrayDiffusion.C?

bwspenc reviewed (commented):
I looked through this and I have no concerns.

Along with our discussion about the non-AD version of this, I assume this will pick up the coupling between array variable entries at a given node in the Jacobian. If so, we should be able to use this for what we want it for.

GiudGiud commented on framework/include/materials/GenericConstant2DArray.h:
```diff
@@ -11,12 +11,13 @@
 #include "Material.h"
 
-class GenericConstant2DArray : public Material
+template <bool is_ad>
```
```suggestion
/**
 * Template for defining a constant array material property, using automatic differentiation or not
 */
template <bool is_ad>
```

GiudGiud commented on framework/include/utils/MooseTypes.h:
```diff
@@ -378,6 +414,7 @@ typedef MooseArray<ADRealVectorValue> ADVariableGradient;
 typedef MooseArray<ADRealTensorValue> ADVariableSecond;
 typedef MooseArray<ADRealVectorValue> ADVectorVariableValue;
 typedef MooseArray<ADRealTensorValue> ADVectorVariableGradient;
+
```
```suggestion
```

GiudGiud commented on framework/src/variables/MooseVariableDataBase.C:
```diff
@@ -620,66 +606,49 @@ MooseVariableDataBase<OutputType>::fetchDoFValues()
-    }
+    dof_values[i].resize(_count);
+    for (const auto j : make_range(_count))
+      dof_values[i](j) = sol(_dof_indices[j * nshapes + i]);
```
so why is there no concern that the vector be an array of lagrange_vec and that for each shape we need to loop from 0 to _count * 3 in this routine?

lindsayad replied:
we are using the `_dof_indices`. We're directly indexing into `_dof_indices` here. We mandate in libMesh that the array dof indexing has major ordering based on count and minor ordering based on shape

lindsayad replied:
We pick this convention because, for example, users can specify different scaling factors for different array variable components. It is easy to extract contiguous submatrices and subvectors corresponding to a given array variable component using the convention we've chosen for multiplication by these potentially disparate scaling factors

GiudGiud reviewed (commented):
last round

lindsayad committed:
Apply suggestions from code review

Better doxygen and class descriptions

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3376883/), step Results summary on 6b9f53e wanted to post the following:

## Framework test summary

Compared against 5a027bf in job [civet.inl.gov/job/3376040](https://civet.inl.gov/job/3376040).

### Added tests

| Test                                                    | Time (s)   |
|---------------------------------------------------------|------------|
| `kernels/array_kernels.test_ad_diffusion_reaction_jac`  | 0.96       |
| `kernels/array_kernels.test_ad_jac`                     | 0.88       |
| `kernels/array_kernels.test_ad`                         | 0.87       |
| `nodalkernels/array-reaction-decay.test/ad_jac`         | 0.84       |
| `kernels/array_kernels.test_ad_diffusion_reaction`      | 0.82       |
| `nodalkernels/array-reaction-decay.test/ad_result_diff` | 0.78       |

### Run time changes

| Test                                              | Base (s)   | Head (s)   | +/-     |
|---------------------------------------------------|------------|------------|---------|
| `fvkernels/mms/advective-outflow.outflow/average` | 28.25      | 8.17       | -71.09% |
## Modules test summary

Compared against 5a027bf in job [civet.inl.gov/job/3376040](https://civet.inl.gov/job/3376040).

No change

GiudGiud reviewed (approved):
looks good to me

moosebuild commented:
Job [Python 3.11](https://civet.inl.gov/job/3376880/) on 6b9f53e : invalidated by @lindsayad

moosebuild commented:
Job [Python 3.11](https://civet.inl.gov/job/3376880/) on 6b9f53e : invalidated by @lindsayad
